scanpipe/pipelines/alpine_packages.py

```diff
@@ -0,0 +1,95 @@
+# You may obtain a copy of the License at: http://apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# Data Generated with ScanCode.io is provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+# ScanCode.io should be considered or used as legal advice. Consult an Attorney
+# for any legal advice.
+#
+# ScanCode.io is a free software code scanning tool from nexB Inc. and others.
+# Visit https://github.com/nexB/scancode.io for support and download.
+
+from scanpipe.pipelines import Pipeline
+from scanpipe.pipes.alpine import (
+    download_or_checkout_aports,
+    get_packages_from_db,
+    prepare_scan_dir,
+    extract_summary_fields,
+)
+from scanpipe.pipes.scancode import run_extractcode, run_scancode
+
+
+class AlpinePackages(Pipeline):
+    """
+    A pipeline to complement missing alpine package data.
+    Downloads and extracts needed information from aports repository and packages source files.
+    """
+
+    @classmethod
+    def steps(cls):
+        return (
+            cls.create_alpine_versions_dict,
+            cls.download_aports_repo,
+            cls.complement_missing_packages_data,
+        )
+
+    scancode_options = ["--copyright", "--summary"]
+
+    def create_alpine_versions_dict(self):
+        """
+        Create a dict, mapping alpine image ids from the database to alpine versions.
+        """
+        self.alpine_versions = {
+            i["image_id"]: i["distro"]["version_id"]
+            for i in self.project.extra_data["images"]
+            if i["distro"]["identifier"] == "alpine"
+        }
+
+    def download_aports_repo(self):
+        """
+        Iterate over every alpine version associated with this project.
+        Download corresponding aports repository branches (alpine versions).
+        """
+        for image_id in self.alpine_versions:
+            download_or_checkout_aports(
+                self.project.tmp_path, self.alpine_versions[image_id]
+            )
+
+    def complement_missing_packages_data(self):
+        """
+        Iterate over alpine packages associated with this project.
+        Checkout aports repository to the corresponding alpine version and commit.
+        Prepare scan target directory, download and extract package's sources.
+        Run scancode and extract missing data (only copyrights for now).
+        Update and save package's missing data to database.
+        """
+        for scan_target_path, scan_result_path, package in get_packages_from_db(
+            self.project
+        ):
+            if (
+                not download_or_checkout_aports(
```

Comment:
Because checking out to a package's alpine version and commit id (provided by the package manager) gives us the best probability of finding correct APKBUILD (for a corresponding package version) and correct patch files, making our scans much more accurate. Also, package versions can change from commit to commit within alpine version branch.

Reply:
> Can we setup a time to discuss all these in a live session?

Yeah, it would be great.